src/libraries/System.Private.CoreLib/src/System/Collections/Generic/HashSetEqualityComparer.cs

```diff
@@ -0,0 +1,78 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Collections.Generic
+{
+    /// <summary>Equality comparer for hashsets of hashsets</summary>
+    internal sealed class HashSetEqualityComparer<T> : IEqualityComparer<HashSet<T>?>
+    {
+        public bool Equals(HashSet<T>? x, HashSet<T>? y)
+        {
+            // If they're the exact same instance, they're equal.
+            if (ReferenceEquals(x, y))
+            {
+                return true;
+            }
+
+            // They're not both null, so if either is null, they're not equal.
+            if (x == null || y == null)
+            {
+                return false;
+            }
+
+            EqualityComparer<T> defaultComparer = EqualityComparer<T>.Default;
+
+            // If both sets use the same comparer, they're equal if they're the same
+            // size and one is a "subset" of the other.
+            if (HashSet<T>.EqualityComparersAreEqual(x, y))
+            {
+                return x.Count == y.Count && y.IsSubsetOfHashSetWithSameComparer(x);
+            }
+
+            // Otherwise, do an O(N^2) match.
+            foreach (T yi in y)
```

Comment:
The code here is obviously wrong due to checking `y` ⊆ `x` only.  I guess no one have been using it.
```C#
var x = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { "foo", "bar" };
var y = new HashSet<string>(StringComparer.Ordinal) { "foo" };
var comparer = HashSet<string>.CreateSetComparer();
Console.WriteLine(comparer.Equals(x, y)); // True (incorrect)
```

Reply:
Yeah, I've never seen anyone use this. I just cleaned up the formatting and deleted some duplicate code. And if anyone is using it, "fixing" it by changing its behavior is likely to cause more grief than it's worth. If we want to, we should open a separate issue; I'm not going to change it here.

Reply:
Apparently the HashSetEqualityComparer bug is tracked by #18931.